### PR TITLE
MGMT-7498: Add missing backwards compatibility for network configuration

### PR DESCRIPTION
**Pull request by mkowalski**
# Assisted Pull Request

## Description

This commit adds a backwards compatibility for setting network
configuration using the following structures

* MachineNetworkCidr
* ClusterNetworkCidr
* ClusterNetworkHostPrefix
* ServiceNetworkCidr

This works in a way that whenever the old data structure is received by
the service, the respective values in both old and new structures are
set internally.

In case when service receives both old and new data structure in the
same request, the old data structure will take precedence.

Follows-up: #2455
Contributes-to: [MGMT-7498](https://issues.redhat.com/browse/MGMT-7498)
Contributes-to: Bug 1999752

## List all the issues related to this PR

- [x] New Feature
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [x] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

### Manual test

+++++++++++++++++++++++++++++
+++ TEST SERVICE NETWORKS +++
+++++++++++++++++++++++++++++

```
$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.service_network_cidr'
"172.30.0.0/16"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.service_networks'
[
  {
    "cidr": "172.30.0.0/16",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b"
  }
]

$ curl --data '{"service_network_cidr":"172.31.0.0/16"}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
[...]

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.service_network_cidr'
"172.31.0.0/16"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.service_networks'
[
  {
    "cidr": "172.31.0.0/16",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b"
  }
]

$ curl --data '{"service_networks":[{"cidr":"172.32.0.0/16"}]}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
[...]

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.service_network_cidr'
"172.32.0.0/16"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.service_networks'
[
  {
    "cidr": "172.32.0.0/16",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b"
  }
]
```

+++++++++++++++++++++++++++++
+++ TEST MACHINE NETWORKS +++
+++++++++++++++++++++++++++++

```
$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.machine_network_cidr'
"192.168.145.0/24"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.machine_networks'
[
  {
    "cidr": "192.168.145.0/24",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b"
  }
]

$ curl --data '{"machine_network_cidr":"192.168.146.0/24"}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
[...]

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.machine_network_cidr'
"192.168.146.0/24"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.machine_networks'
[
  {
    "cidr": "192.168.146.0/24",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b"
  }
]

$ curl --data '{"machine_networks":[{"cidr":"192.168.147.0/24"}]}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
[...]

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.machine_network_cidr'
"192.168.147.0/24"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.machine_networks'
[
  {
    "cidr": "192.168.147.0/24",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b"
  }
]
```

+++++++++++++++++++++++++++++
+++ TEST CLUSTER NETWORKS +++
+++++++++++++++++++++++++++++

```
$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_network_cidr'
"10.128.0.0/14"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_network_host_prefix'
23

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_networks'
[
  {
    "cidr": "10.128.0.0/14",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b",
    "host_prefix": 23
  }
]

$ curl --data '{"cluster_network_cidr":"10.124.0.0/14"}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
[...]

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_network_cidr'
"10.124.0.0/14"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_network_host_prefix'
23

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_networks'
[
  {
    "cidr": "10.124.0.0/14",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b",
    "host_prefix": 23
  }
]

$ curl --data '{"cluster_network_host_prefix":25}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
[...]

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_network_cidr'
"10.124.0.0/14"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_network_host_prefix'
25

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_networks'
[
  {
    "cidr": "10.124.0.0/14",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b",
    "host_prefix": 25
  }
]

$ curl --data '{"cluster_networks":[{"cidr":"10.132.0.0/14","host_prefix":24}]}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
[...]

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_network_cidr'
"10.132.0.0/14"

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_network_host_prefix'
24

$ curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.cluster_networks'
[
  {
    "cidr": "10.132.0.0/14",
    "cluster_id": "82b8c284-8024-4a30-9bef-4b5a8180061b",
    "host_prefix": 24
  }
]
```

## Assignees

/cc @YuviGold 
/cc @flaper87 
/cc @celebdor 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by mkowalski** [comment]
/retitle [WIP] MGMT-7498: Add missing backwards compatibility for network configuration

**Comment by mkowalski** [comment]
/retitle MGMT-7498: Add missing backwards compatibility for network configuration
/hold cancel

**Comment by mkowalski** [comment]
/hold

Manual test failed

**Comment by mkowalski** [comment]
Code built as `quay.io/mkowalski/assisted-service:backwards-compatibility-reversed-order-blabla-3`. Another attempt of the manual test from "How was this code tested?" would be appreciated.

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -2451,9 +2456,14 @@ func (b *bareMetalInventory) updateNetworks(db *gorm.DB, params installer.Update
 				machineNetworkCidr = string(cluster.MachineNetworks[index].Cidr)
 			}
 
+			serviceNetworkCidr := ""
+			if len(cluster.ServiceNetworks) > index {
```
Comment (mkowalski) [inline internal/bminventory/inventory.go:2460]: That's a piggyback removing the artificial limitation of number of service network equals to the number of cluster networks. Without this, if there are more cluster networks than service networks, the code will panic.

File: internal/bminventory/inventory_test.go [hunk 2]
```diff
@@ -5038,6 +5038,25 @@ var _ = Describe("cluster", func() {
 					Expect(actual.Payload.MachineNetworks).To(BeEmpty())
 				})
 
+				It("Empty networks (old API) - valid", func() {
```
Comment (flaper87) [inline internal/bminventory/inventory_test.go:5041]:
```suggestion
				It("Empty networks v1 - valid", func() {
```

File: internal/bminventory/inventory_test.go [hunk 3]
```diff
@@ -5020,7 +5020,7 @@ var _ = Describe("cluster", func() {
 					validateNetworkConfiguration(actual.Payload, &clusterNetworks, &serviceNetworks, &machineNetworks)
 				})
 
-				It("Empty networks - valid", func() {
+				It("Empty networks (new API) - valid", func() {
```
Comment (flaper87) [inline internal/bminventory/inventory_test.go:5023]:
```suggestion
				It("Empty networks dual-stack - valid", func() {
```

Maybe we should just call it `v2` (even if it's not under the v2 path)

**Review by flaper87 — CHANGES_REQUESTED** [review]
looks good, one comment on the `createNetworkParamsCompatibilityPropagation` implementation plus some name suggestions for the tests.

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -2510,6 +2520,100 @@ func (b *bareMetalInventory) updateNetworkTables(db *gorm.DB, cluster *common.Cl
...
+//   * ClusterNetworkHostPrefix
+//   * ServiceNetworkCidr
+//
+// Please note those will take precedence over the more complex introduced as part of dual-stack
+// and multi-network support, i.e.
+//
+//   * MachineNetworks
+//   * ClusterNetworks
+//   * ServiceNetworks
+func createNetworkParamsCompatibilityPropagation(params installer.V2RegisterClusterParams) {
```
[the remark this answers is not in the document]
Comment (flaper87) [reply internal/bminventory/inventory.go:2537]: I dug a bit more into this and noticed there's no `MachineNetworkCidr` in the [`CreateClusterParams` struct](https://github.com/openshift/assisted-service/blob/3ed387eda54e012ac597bf48df8c9286b35d13fc/models/cluster_create_params.go#L21).

I guess there's nothing missing in this PR, then.

**Review by flaper87 — APPROVED** [review]

**Pull request merged**